### DISCUSSION TITLE
Provide a clean way to strip the token from an URL

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#8914] [Ideas] [answered]
### Description
Currently when a third party service redirects to a Craft CMS project with a `token` parameter in the URL it's sometimes difficult to handle it well, because all generated links in the system will attach that token again and again. 
But there are some routes that will break in that case.
It would be awesome to have a way to strip that token from generated URLs, currently that's not possible if you want to use inbuild functions because `UrlHelper::_createUrl` is private and passing the `$addToken` param all the way there from `$element->getUrl()` seems over the top.

I'm not that sure what the best solution would be, my current (ugly) work around is to temporarily change the token param so the function does not recognize it.
Maybe a regex that strips it?
I'm open for a "we will refactor everything in Craft 4, just change the `GeneralConfig->tokenParam` for the time being" as well.
Maybe by using Classes instead of static functions (as Stringify does)

```php
$url = new Url();
$url->path = 'foo/bar';
$url->params = ['foo' => 'bar'];
$url->scheme = '...';
$url->getUrl(); // or via __toString();
```
or something to have full control over all params instead of many parameters in function.
So you can do `{{ entry.getUrl() }}` as you are used to and additionally `{{ entry.getUrl().params([]) }}`

So in the end your `_createUrl` would always return an object with a `__toString` instead of a string directly. I think this is much cleaner.

**Edit**
Forget what I wrote about changing the token param, Craft will just attach it as a different parameter in that case

**Comment by brandonkelly** [comment, marked as answer]
You could use [UrlHelper::removeParam()](https://docs.craftcms.com/api/v3/craft-helpers-urlhelper.html#method-removeparam) for this.

```php
$url = UrlHelper::removeParam($element->getUrl(), Craft::$app->config->general->tokenParam);
```

The `Url` class is a cool idea, but not sure it would really help outside of a major release, since we wouldn’t want to change the `Element::getUrl()` signature. So you’d end up having to do something like

```php
$url = (new Url($element->getUrl())->withoutParam(Craft::$app->config->general->tokenParam);
```

Slightly nicer syntax maybe, but not saving you anything. Will leave this open as something to consider down the road though.

**Comment by Anubarak** [comment]
Thank you, I totally overlooked the `removeParam` function. Yeah my idea to do it via class was for the next major release. Of course it won't be used that much but I guess your code will become a little bit cleaned up as well (lesser parameters for each function)
